### PR TITLE
Topic event type improvements (revision)

### PR DESCRIPTION
**Pull request by telephon**
This is a revision of #2227, and is a final result of the discussion.

### CODE REVIEW

File: testsuite/classlibrary/TestEvent.sc [hunk 1]
```diff
@@ -34,6 +34,50 @@ TestEvent : UnitTest {
 
 	}
 
+	test_parent_proto {
```
Comment (brianlheim) [inline testsuite/classlibrary/TestEvent.sc]: if this test really needs to be this complicated, it would help to have an ASCII diagram or some other sort of description in the code, and not have to read the test asserts to reconstruct it

File: HelpSource/Classes/Event.schelp [hunk 2]
```diff
@@ -36,31 +36,91 @@ Returns an empty event with link::#.defaultParentEvent:: as parent.
...
+discussion::
 code::
-Event.addEventType(\happyEvent, { ("I am so happy to be silent sometimes, says" + ~who).postln; });
+Event.addEventType(\happyEvent, { "I am so happy to be silent sometimes, says %.\n".postf(~who) });
+(type: \happyEvent, who: "Alice").play;
+
+// using protoEvent
+(
+Event.addEventType(\happyEvent, {
+	"I am so happy to be silent sometimes, says %.\n".postf(~who)
```
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: maybe keep the indentation the same as in the previous (parent event) definition?

File: HelpSource/Classes/Event.schelp [hunk 3]
```diff
@@ -36,31 +36,91 @@ Returns an empty event with link::#.defaultParentEvent:: as parent.
...
+)
+
+(type: \happyEvent).play; // use default
+(type: \happyEvent, who: "Eve").play; // overrride default
+
+// in a Pbind:
 Pbind(\type, \happyEvent, \who, Prand(["Alice", "Bob", "Eve"], inf), \dur, Pwhite(0.1, 1.0, inf)).play;
 
-// To a certain degree, it is possible to reuse some of another event type's functionality:
+// protoEvent for indirect calls:
```
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: at this point, protoEvent is not mentioned in the document, and here it is used implicitly. Makes it hard to follow.

File: HelpSource/Classes/Event.schelp [hunk 4]
```diff
@@ -36,31 +36,91 @@ Returns an empty event with link::#.defaultParentEvent:: as parent.
 method::silent
 Returns an event that describes a pause of strong::dur:: duration.
 
+
+
 method::addEventType
-Event types define alternate play functions that are selected by the value of strong::~type::.
+Event types define alternate play functions and parent events that are selected by the value of strong::~type::.
```
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: so, addEventType adds key/value pairs to parent? maybe this could be written here.

File: HelpSource/Classes/Event.schelp [hunk 5]
```diff
@@ -36,31 +36,91 @@ Returns an empty event with link::#.defaultParentEvent:: as parent.
...
 	~type = \note; // now set type to a different one
 	currentEnvironment.play;
 });
+
 Pbind(\type, \happyEvent, \degree, Pseq([0, 1, 2, 3, 4, 4, 5, 5, 5, 5, 4, 2, 3, 2, 3, 1], inf), \dur, Pwhite(0.1, 1.0, inf)).play;
 );
 
 ::
 
+The event types and parent events are stored in a dictionary of the Event class, namely in code::partialEvents.playerEvent:::
```
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: should be ```Event.partialEvents.playerEvent```.
Actually, it'd be nice to have methods such as 

```
listParentTypes
listEventTypes
```

to give users convenient access to possible values for arguments of the `play()` method.

File: HelpSource/Classes/Event.schelp [hunk 6]
```diff
@@ -36,31 +36,91 @@ Returns an empty event with link::#.defaultParentEvent:: as parent.
...
+
+code::
+// eventTypes
+Event.partialEvents.playerEvent.eventTypes
+// parentTypes
+Event.partialEvents.playerEvent.parentTypes
+::
```
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: so, we have `Proto` and `Parent`, yet there are `eventTypes` and `parentTypes`? I am totally lost by now.

File: HelpSource/Classes/Event.schelp [hunk 7]
```diff
@@ -36,31 +36,91 @@ Returns an empty event with link::#.defaultParentEvent:: as parent.
 method::silent
```
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: + "related" should include Classes/Environment and Classes/Dictionary
+ the short description in line 18 should read:

An Event specifies an action to be taken in response to a link::#-play:: message. Its key/value pairs specify the parameters of that action. Event inherits most of its methods from its superclasses, especially from link::Classes/Dictionary::. __+ Something explaining the usage and differences of code::parent:: and code::proto::.__

File: SCClassLibrary/Common/Collections/Event.sc [hunk 8]
```diff
@@ -414,21 +422,30 @@ Event : Environment {
 				type: \note,
 
 				play: #{
-					var tempo, server;
+					var tempo, server, eventTypes, parentType;
+
+					parentType = ~parentTypes[~type];
+					parentType !? { currentEnvironment.parent = parentType };
```
Comment (LFSaw) [inline SCClassLibrary/Common/Collections/Event.sc]: which type of if.isNil statement do you want to use?

File: SCClassLibrary/Common/Collections/Event.sc [hunk 9]
```diff
@@ -414,21 +422,30 @@ Event : Environment {
...
 
 					~finish.value(currentEnvironment);
 
-					server = ~server ?? { Server.default };
+					server = ~server ? Server.default;
 
 					tempo = ~tempo;
-					if (tempo.notNil) {
-						thisThread.clock.tempo = tempo;
+					if(tempo.notNil) {
```
Comment (LFSaw) [inline SCClassLibrary/Common/Collections/Event.sc]: which type of if.isNil statement do you want to use?

File: HelpSource/Classes/Event.schelp (hunk 3 above)
Comment (telephon) [inline HelpSource/Classes/Event.schelp]: it's wrong (leftover!) should be parent, of course.

File: HelpSource/Classes/Event.schelp (hunk 5 above)
Comment (telephon) [inline HelpSource/Classes/Event.schelp]: Like this?
```
*listParentTypes {
		Event.partialEvents.playerEvent.parentTypes.keys
	}
	*listEventTypes {
		Event.partialEvents.playerEvent.eventTypes.keys
	}
```

File: HelpSource/Classes/Event.schelp (hunk 6 above)
Comment (telephon) [inline HelpSource/Classes/Event.schelp]: yes, every event has a proto and a parent. But in the event class, you can access play functions (so called event types) and their respective default parameters (so called parent types).

File: testsuite/classlibrary/TestEvent.sc (hunk 1 above)
Comment (telephon) [inline testsuite/classlibrary/TestEvent.sc]: let's see, maybe I can do one. There is a graphics in IdentityDictionary help.

File: HelpSource/Classes/Event.schelp (hunk 5 above)
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: yep, but with `^returns` :)

File: HelpSource/Classes/Event.schelp (hunk 6 above)
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: this should go into the helpfile...

File: HelpSource/Classes/Event.schelp (hunk 5 above)
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: also, maybe all defined types for an Event instance?

File: HelpSource/Classes/Event.schelp (hunk 7 above)
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: now you refer to `IdentityDictionary` for parent and proto, so this needs to go into related (instead of `Dictionary`?)

File: HelpSource/Classes/Event.schelp [hunk 10]
```diff
@@ -39,7 +39,7 @@ Returns an event that describes a pause of strong::dur:: duration.
 
 
 method::addEventType
-Event types define alternate play functions and parent events that are selected by the value of strong::~type::.
+Event types define alternate play functions and parent events that are selected by the value of strong::~type::. Practically, this allows to augment the default parent event with key value pairs for this type only.
```
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: ".... selected by the value stored at its strong::type:: key..." ?

Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: ".... selected by the value stored at its strong::type:: key..." ?

File: HelpSource/Classes/Event.schelp (hunk 2 above)
Comment (LFSaw) [inline HelpSource/Classes/Event.schelp]: fixed.

**Comment by telephon** [comment]
I have no time to revise the whole helpfile soon. Maybe that could be another PR?

**Comment by telephon** [comment]
@LFSaw better?

File: SCClassLibrary/Common/Collections/Event.sc (hunk 8 above)
Comment (LFSaw) [inline SCClassLibrary/Common/Collections/Event.sc]: fixed.

File: testsuite/classlibrary/TestEvent.sc (hunk 1 above)
Comment (LFSaw) [inline testsuite/classlibrary/TestEvent.sc]: fixed.

**Comment by LFSaw** [comment]
> I have no time to revise the whole helpfile soon. Maybe that could be another PR?

sure... thing is that, as the history in this PR and in #2227 indicates, the mechanisms of `Event` and `Event.default` in particular are by far not self-explanatory. Of course, there's the help file already and the (extensive) collection of tutorials around patterns, yet, a quick overview on the workings, some kind of cheat-sheet (a concept I closely associate with class help files) would be very good to have. If I'd be more knowledgeable in the Pattern world, I'd start improve it myself, but, hmmm... I'm not :/

**Comment by snappizz** [comment]
i don't get what this PR did, and i'd appreciate if someone could summarize it for the changelog. are there any API changes?
